src/features/go-to-symbol.ts

```diff
@@ -0,0 +1,319 @@
+import * as ElmToAst from './shared/elm-to-ast'
+import * as ElmSyntax from './shared/elm-to-ast/elm-syntax'
+
+// This finds all top-level declarations (type, type alias, values, functions, ports)
+// with the following caveats:
+// - Odd spacing or inline comments can cause declarations to be missed.
+// - False positives might be found in multiline comments or multiline strings.
+// But it’s a very fast way of computing an important feature!
+// If there are a couple of false positives you can jump to, that’s not the end of the world.
+// For example, if you comment out some code using multiline comments, you’ll still be able
+// to jump to it, which is a bit unexpected but not super weird.
+// Parts come from here: https://github.com/rtfeldman/node-test-runner/blob/eedf853fc9b45afd73a0db72decebdb856a69771/lib/Parser.js#L229-L234
+const topLevelDeclarationRegex = /\n|(?<=^type(?: +alias)? +)\p{Lu}[_\d\p{L}]*|^\p{Ll}[_\d\p{L}]*(?=.*=$)|^(?<=port +)\p{Ll}[_\d\p{L}]*(?= *:)/gmu
+
+type Fallback = {
+  fsPath: string
+  symbols: vscode.DocumentSymbol[]
+}
+
+export const feature: Feature = ({ context }) => {
+  let fallback: Fallback | undefined = undefined
+
+  context.subscriptions.push(
+    vscode.languages.registerDocumentSymbolProvider('elm', {
+      async provideDocumentSymbols(doc: vscode.TextDocument, token: vscode.CancellationToken) {
+        // Allow user to disable this feature
+        const isEnabled: boolean = vscode.workspace.getConfiguration('elmLand').feature.goToSymbol
+        if (!isEnabled) return
+
+        const start = Date.now()
+        const text = doc.getText()
+        const ast = await ElmToAst.run(text, token)
+
+        let symbols: vscode.DocumentSymbol[] = []
+        if (ast) {
+          symbols = ast.declarations.map(declarationToDocumentSymbol)
+          fallback = {
+            fsPath: doc.uri.fsPath,
+            symbols,
+          }
+        } else if (fallback !== undefined && doc.uri.fsPath === fallback.fsPath) {
+          // When you start editing code, it won’t have correct syntax straight away,
+          // but VSCode will re-run this. If you have the Outline panel open in the sidebar,
+          // it’s quite distracting if we return an empty list here – it will flash
+          // between “no symbols” and all the symbols. So returning the symbols from last
+          // time we got any improves the UX a little. Note: If you remove all text in the file,
+          // the Outline view shows old stuff that isn’t available – until the file becomes
+          // syntactically valid again – but I think it’s fine.
+          symbols = fallback.symbols
+        }
+        
+        console.info('provideDocumentSymbol', `${symbols.length} results in ${Date.now() - start}ms`)
+        return symbols
+      }
+    })
+  )
+
+  context.subscriptions.push(
+    vscode.languages.registerWorkspaceSymbolProvider({
+      async provideWorkspaceSymbols(query: string, token: vscode.CancellationToken) {
```

Comment:
TODO: This should probably be a separate feature, because if one is slow then you can disable it and you can still use the other.

Reply:
Did that